llms/llama/convert.py

```diff
@@ -37,14 +36,15 @@ def unshard(k, v):
             axis = 1
         else:
             raise ValueError("Invalid weight name")
-        return np.concatenate(v, axis=axis)
+        return mx.concatenate(v, axis=axis)
 
     torch_files = glob.glob(str(model_path / "consolidated.*.pth"))
     weights = collections.defaultdict(list)
     for wf in torch_files:
         state = torch.load(wf, map_location=torch.device("cpu"))
         for k, v in state.items():
-            v = v.to(torch.float16).numpy()
+            v = mx.array(v.to(torch.float32).numpy(), mx.bfloat16)
```

Comment:
The only thing I'm wondering about this change is that it will default the model inference to `mx.bfloat16` which might be a lot slower than `mx.float16`. If that's the case (I haven't tested it yet) I might suggest that we give the option of choosing the output type.

So it could be an argument `--type` which is any of `mx.float16`, `mx.bfloat` and `mx.float32`? Wdyt?

Reply:
I'll measure performance comparing float16 and bfloat16. I would probably add the --type parameter to the inference script if needed at all. But lets not optimizer premature and measure performance first.

Reply:
Found the time to do some measurements on inference and it looks quite encouraging. 

I run basically: 
```
    for dtype in [mx.float16, mx.bfloat16, mx.float32]:
        model.update(tree_map(lambda x: x.astype(dtype), model.parameters()))
        repeat = 3
        print("{}: best out of {}: {:.3f}s".format(dtype, repeat, min(timeit.repeat("generate(args)", number=1, repeat=repeat, globals=globals()))), file=sys.stderr)
```
for the `llama-2-13b-chat` with default parameters for generation. 

Results from my M3 Max Mac:
```
mlx.core.float16: best out of 3: 8.660s
mlx.core.bfloat16: best out of 3: 8.520s
mlx.core.float32: best out of 3: 192.604s
```

bfloat16 performs actually a tiny bit faster than float16, but I think that's not statistically significant. For comparison, float32 is roughly 200% slower. 

So from my point of view there is no need for a `--type` parameter.

Reply:
Wow 200% slower for fp32. That's bad.. it must not fit well on your machine.  I'm cool with defaulting this to `bfloat16`. Do you want to submit the PR for review?

Reply:
That is probably true, 200% percent is unexpected. Probably the model is too big in float32, but I did not look into it. I did some general matrix multiplication timing, where I see around 3% to 5% slowdown for float32, depending on the matrix size. 
Sure, we can review/merge this PR and create follow-up PRs for other models.

Reply:
forgive me if I'm wrong, but isn't `192.604` more like 24x to `8.520` than 200%?

Reply:
just from the number alone I thought I would've been difference between ANE or GPU, until I saw [this](https://github.com/ml-explore/mlx/issues/18)

Reply:
> forgive me if I'm wrong, but isn't `192.604` more like 24x to `8.520` than 200%?

@Willian-Zhang I don't remember what my computation was, but you are probably right. We may just agree that it's just a big difference. 

As awni already pointed out, this is unexpected, and I do not see this kind of performance degradation for float32 in general. If you experience performance issues with float32, it might be worth opening a separate issue for that topic. 
I understood that there are still some [performance improvements underway](https://github.com/ml-explore/mlx/issues/243#issuecomment-1868376661), so I will hold back with further performance benchmarks until the next version of mlx.

Reply:
I'm getting some different numbers on an M1 mac
```
f16: 350 token/s
bf16: 100 token/s
f32: 137 token/s
```

Reply:
It's most likely not memory constrained from Page-In out status, however I'm not sure how I would be able to tell exactly.

Reply:
it seems bf16 perform less to f32 in this case, and AFAIK M1 does not support bf16 natively. I'm not sure if its stored f32 internally and somehow converting bf16 f32 back and forth or entirely simulated with f32 under the hood.

Reply:
Those timings are surprising! I would not expect (even with native support) for FP16 to be more than 2x faster than fp32.

Can you say more about what you ran in that benchmark?
Which LLM,  your OS, etc.

Reply:
mistral 7B, 
14.2.1 (23C71), 
m1 ultra, 20c, 64c, 32c, 128G

Reply:
What did you use to benchmark? I just ran mistral 7B in FP16 on an m2ultra with 192 GB and I see more like 30 tokens-per-secon.

Reply:
a modified version of the lora.py, with some custom data varying around 2048 token per datapoint.

Reply:
@awni 
<img width="521" alt="image" src="https://github.com/ml-explore/mlx-examples/assets/5165840/3e7432e6-05e8-4338-aaf5-a828583c9995">
may I ask which one did you use?

Reply:
> bs=1
I felt padding on mlx currently without careful thought on parallelism would result in performance impact, so larger context Len, lowered bs to 1

Reply:
The one with 76-core GPU.

Reply:
> The one with 76-core GPU.

well that 's surprising

Reply:
> I felt padding on mlx currently without careful thought on parallelism would result in performance impact, so larger context Len, lowered bs to 1

Fair enough. The padding + batching strategy is very naive in that example. It should try to put similar length samples in the same batch.

Reply:
> well that 's surprising

No I was timing generation (which is very different from training). 300 tps for training is sensible, even on the slow side.

Reply:
> No I was timing generation (which is very different from training).

Oh. okey now it make sense.